extensions/poetry/main.ts

```diff
@@ -0,0 +1,53 @@
+import { PhylumApi } from "phylum"
+import { parse } from "https://deno.land/std/flags/mod.ts"
+
+// Add packages to the lockfile, and upload it to Phylum for analysis.
+// Return `true` if the packages pass the project's thresholds.
+// Return `false` if the packages don't pass the thresholds, or the processing
+// job is not finished.
+async function poetryCheck(args: string[]) {
+    console.log("Updating package lock…")
+
+    let process = Deno.run({ cmd: ["poetry", "add", '--lock', ...args] })
+    await process.status()
+    await process.close()
+
+    console.log("Package lock updated.\n")
+    console.log("Analyzing packages…")
+
+    const jobId = await PhylumApi.analyze("./poetry.lock")
+    const jobStatus = await PhylumApi.getJobStatus(jobId)
+
+    if (jobStatus.pass && jobStatus.status === "complete") {
+        console.log("All packages pass project thresholds.\n")
+        return true
+    } else if (jobStatus.pass) {
+        console.warn("Unknown packages were submitted for analysis, please check again later.")
+        return false
+    } else {
+        let packages = pkgs.join("', '")
+        console.error(`Installing '${packages}' caused threshold failure.`)
+        return false
+    }
+}
+
+// If the subcommand is `add`, update the lockfile and process it through
+// Phylum. Otherwise, pass the arguments through to `poetry`.
```

Comment:
Good to know -- are there any other packages that write to the lockfile that we may have to support?

Reply:
The belief is that `poetry` is the only package/tool that modifies the `poetry.lock` lockfile and therefore the only one to support.

Reply:
Sorry, I meant _subcommands_ rather than _packages_. 😅

Reply:
Ah. A review of the `poetry` subcommands resulted in just two that **directly** add dependencies to the lockfile: `add` and `update`. There are other subcommands that touch/modify the lockfile, but they do so **indirectly** (i.e., not because a user supplied a specific package). These commands are: `install` and `lock`. There is also the possibility that `remove` will update the lockfile in a way that not only removes the specified package, but re-resolves the dependencies, resulting in a lockfile that has new/different packages than from before the command was run. Documentation for these commands is [here](https://python-poetry.org/docs/cli) and I'd be more than willing to try any specific scenario for you (or screen share to see it together).